server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepositoryTests.java

```diff
@@ -161,28 +168,45 @@ public void testOverwriteSnapshotInfoBlob() throws Exception {
             final SnapshotId snapshotId = new SnapshotId("foo", UUIDs.randomBase64UUID());
             TestFutureUtils.<RepositoryData, Exception>get(f ->
                 // We try to write another snap- blob for "foo" in the next generation. It fails because the content differs.
-                repository.finalizeSnapshot(ShardGenerations.EMPTY, RepositoryData.EMPTY_REPO_GEN, Metadata.EMPTY_METADATA,
-                    new SnapshotInfo(snapshotId, Collections.emptyList(),
-                        0L, null, 1L, 5, Collections.emptyList(), true),
-                    Version.CURRENT, Function.identity(), f));
+                repository.finalizeSnapshot(
+                    ShardGenerations.EMPTY,
+                    RepositoryData.EMPTY_REPO_GEN,
+                    Metadata.EMPTY_METADATA,
+                    new SnapshotInfo(snapshotId, Collections.emptyList(), 0L, null, 1L, 5, Collections.emptyList(), true),
+                    Version.CURRENT,
+                    Function.identity(),
+                    f
+                ));
 
             // We try to write another snap- blob for "foo" in the next generation. It fails because the content differs.
-            assertThatThrownBy(
-                () -> TestFutureUtils.<RepositoryData, Exception>get(f ->
-                    repository.finalizeSnapshot(ShardGenerations.EMPTY, 0L, Metadata.EMPTY_METADATA,
-                        new SnapshotInfo(snapshotId, Collections.emptyList(),
-                            0L, null, 1L, 6, Collections.emptyList(), true),
-                        Version.CURRENT, Function.identity(), f))
-            ).isInstanceOf(AssertionError.class)
-                .hasMessage("\nExpected: <6>\n     but: was <5>");
+            FutureActionListener<RepositoryData, RepositoryData> fut = FutureActionListener.newInstance();
+            repository.finalizeSnapshot(
+                ShardGenerations.EMPTY,
+                0L,
+                Metadata.EMPTY_METADATA,
+                new SnapshotInfo(snapshotId, Collections.emptyList(), 0L, null, 1L, 6, Collections.emptyList(), true),
+                Version.CURRENT,
+                Function.identity(),
+                fut
+            );
+            assertThat(fut).failsWithin(5, TimeUnit.SECONDS)
+                .withThrowableOfType(ExecutionException.class)
+                .havingCause()
+                    .isExactlyInstanceOf(AssertionError.class)
+                    .withMessage("\nExpected: <6>\n     but: was <5>");
```

Comment:
The only reason this test ever passed was because the test layer mocks the threadPool to use a `SameThreadExecutorService`, which means `executor.execute()` propagated raised errors. 

`ActionRunnable` otherwise only forwards `Exception` to the listenable, not `Error`.